File: app/src/main/java/com/zack/enderplan/model/bean/Plan.java
package com.zack.enderplan.model.bean;

import android.os.Parcel;
import android.os.Parcelable;

import com.zack.enderplan.util.Constant;

public class Plan implements Parcelable {

    private static final int PLAN_STAR_STATUS_NOT_STARRED = 0;
    private static final int PLAN_STAR_STATUS_STARRED = 1;

    private String mPlanCode;
    private String mContent;
    private String mTypeCode;
    private long mCreationTime;
    private long mDeadline;
    private long mCompletionTime;
    private int mStarStatus;
    private long mReminderTime;

    public Plan(String planCode) {
        this.mPlanCode = planCode;
        this.mContent = null;
        this.mTypeCode = null;
        this.mCreationTime = Constant.UNDEFINED_TIME;
        this.mDeadline = Constant.UNDEFINED_TIME;
        this.mCompletionTime = Constant.UNDEFINED_TIME;
        this.mStarStatus = PLAN_STAR_STATUS_NOT_STARRED;
        this.mReminderTime = Constant.UNDEFINED_TIME;
    }

    public Plan(String planCode, String content, String typeCode, long creationTime, long deadline,
                long completionTime, int starStatus, long reminderTime) {
        this.mPlanCode = planCode;
        this.mContent = content;
        this.mTypeCode = typeCode;
        this.mCreationTime = creationTime;
        this.mDeadline = deadline;
        this.mCompletionTime = completionTime;
        this.mStarStatus = starStatus;
        this.mReminderTime = reminderTime;
    }

    protected Plan(Parcel in) {
        mPlanCode = in.readString();
        mContent = in.readString();
        mTypeCode = in.readString();
        mCreationTime = in.readLong();
        mDeadline = in.readLong();
        mCompletionTime = in.readLong();
        mStarStatus = in.readInt();
        mReminderTime = in.readLong();
    }

    public static final Creator<Plan> CREATOR = new Creator<Plan>() {
        @Override
        public Plan createFromParcel(Parcel in) {
            return new Plan(in);
        }

        @Override
        public Plan[] newArray(int size) {
            return new Plan[size];
        }
    };

    public String getPlanCode() {
        return mPlanCode;
    }

    public void setPlanCode(String planCode) {
        this.mPlanCode = planCode;
    }

    public String getContent() {
        return mContent;
    }

    public void setContent(String content) {
        this.mContent = content;
    }

    public String getTypeCode() {
        return mTypeCode;
    }

    public void setTypeCode(String typeCode) {
        this.mTypeCode = typeCode;
    }

    public long getCreationTime() {
        return mCreationTime;
    }

    public void setCreationTime(long creationTime) {
        this.mCreationTime = creationTime;
    }

    public long getDeadline() {
        return mDeadline;
    }

    public void setDeadline(long deadline) {
        this.mDeadline = deadline;
    }

    public long getCompletionTime() {
        return mCompletionTime;
    }

    public void setCompletionTime(long completionTime) {
        this.mCompletionTime = completionTime;
    }

    public int getStarStatus() {
        return mStarStatus;
    }

    public void setStarStatus(int starStatus) {
        this.mStarStatus = starStatus;
    }

    public long getReminderTime() {
        return mReminderTime;
    }

    public void setReminderTime(long reminderTime) {
        this.mReminderTime = reminderTime;
    }

    public boolean hasDeadline() {
        return mDeadline != Constant.UNDEFINED_TIME;
    }

    public boolean isCompleted() {
        return mCompletionTime != Constant.UNDEFINED_TIME;
    }

    public boolean isStarred() {
        return mStarStatus == PLAN_STAR_STATUS_STARRED;
    }

    public boolean hasReminder() {
        return mReminderTime != Constant.UNDEFINED_TIME;
    }

    public void invertStarStatus() {
        setStarStatus(isStarred() ? PLAN_STAR_STATUS_NOT_STARRED : PLAN_STAR_STATUS_STARRED);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(mPlanCode);
        dest.writeString(mContent);
        dest.writeString(mTypeCode);
        dest.writeLong(mCreationTime);
        dest.writeLong(mDeadline);
        dest.writeLong(mCompletionTime);
        dest.writeInt(mStarStatus);
        dest.writeLong(mReminderTime);
    }
}

File: app/src/main/kotlin/me/imzack/app/end/model/bean/Plan.kt
package me.imzack.app.end.model.bean

import android.os.Parcel
import android.os.Parcelable
import me.imzack.app.end.model.DataManager
import me.imzack.app.end.util.CommonUtil

data class Plan(
        val code: String = CommonUtil.makeCode(),
        var content: String = "",
        var typeCode: String = DataManager.getType(0).code,
        var creationTime: Long = System.currentTimeMillis(),
        var deadline: Long = 0L,
        var completionTime: Long = 0L,
        var starStatus: Int = STAR_STATUS_NOT_STARRED,
        var reminderTime: Long = 0L
) : Parcelable {

    companion object {

        private val STAR_STATUS_NOT_STARRED = 0
        private val STAR_STATUS_STARRED = 1

        @Suppress("unused")
        @JvmField
        val CREATOR = object : Parcelable.Creator<Plan> {

            override fun createFromParcel(src: Parcel) = Plan(src)

            override fun newArray(size: Int): Array<Plan?> = arrayOfNulls(size)
        }
    }

    constructor(src: Parcel) : this(
            src.readString(),
            src.readString(),
            src.readString(),
            src.readLong(),
            src.readLong(),
            src.readLong(),
            src.readInt(),
            src.readLong()
    )

    val hasDeadline
        get() = deadline != 0L

    val isCompleted
        get() = completionTime != 0L

    val isStarred
        get() = starStatus == STAR_STATUS_STARRED

    val hasReminder
        get() = reminderTime != 0L

    fun invertStarStatus() {
        starStatus = if (isStarred) STAR_STATUS_NOT_STARRED else STAR_STATUS_STARRED
    }

    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(code)
        dest.writeString(content)
        dest.writeString(typeCode)
        dest.writeLong(creationTime)
        dest.writeLong(deadline)
        dest.writeLong(completionTime)
        dest.writeInt(starStatus)
        dest.writeLong(reminderTime)
    }
}
